src/wallet/scriptpubkeyman.cpp

```diff
@@ -149,7 +191,36 @@ bool LegacyScriptPubKeyMan::CanGetAddresses(bool internal)
 
 bool LegacyScriptPubKeyMan::Upgrade(int prev_version, int new_version, std::string& error)
 {
-    return false;
+    LOCK(cs_KeyStore);
+    error = "";
+    bool hd_upgrade = false;
+    bool split_upgrade = false;
+    if (CanSupportFeature(FEATURE_HD) && !IsHDEnabled()) {
+        WalletLogPrintf("Upgrading wallet to HD\n");
+        SetMinVersion(FEATURE_HD);
+
+        // generate a new master key
+        SetHDSeed(GenerateNewSeed());
+        hd_upgrade = true;
+    }
+    // Upgrade to HD chain split if necessary
+    if (CanSupportFeature(FEATURE_HD_SPLIT) && hdChain.nVersion < 2 /* VERSION_HD_CHAIN_SPLIT */) {
```

Comment:
Referencing `VERSION_HD_CHAIN_SPLIT` directly requires including walletdb.h which is a circular dependency.

The logic is slightly different here because without it, it would end up marking the newly generated keys as pre-split on later runs on the same LegacyScriptPubKeyMan. This is an issue because for a LegacyScriptPubKeyMan, Upgrade() will be called 6 times (once for each address type and internalness). If you were to run `-upgradewallet` multiple times on master, you would see the same problem.